rxfeedback/src/main/java/org/notests/rxfeedback/Feedbacks.kt

```diff
@@ -204,6 +164,186 @@ fun <State, Query, Event> reactSetSafe(
         }
 
 
+/**
+ * The purpose of QueryLifetimeTracking is to activate, deactivate,
+ * as well as update already activated effects with the new Query that was returned.
+ */
+private class QueryLifetimeTracking<Query, QueryID, Event>(
+        val effects: (intialQuery: Query, state: Observable<Query>) -> Observable<Event>,
+        val scheduler: Scheduler,
+        val emitter: Emitter<Event>) {
+
+    /**
+     * Used to make sure that we only get events from the effects that are still active.
+     * Prevents concurrency issues caused by reactivation of the same effect (same QueryID)
+     */
+    private class LifetimeToken
+
+    /**
+     * Used to track the effect that was started and update it with the new value of the Query.
+     */
+    private data class QueryLifetime<Query>(val subscription: Disposable,
+                                    val lifetimeIdentifier: LifetimeToken,
+                                    val latestQuery: BehaviorSubject<Query>)
+
+    /**
+     * State controls the effects by QueryID.
+     */
+    private data class State<QueryID, Query>(var isDisposed: Boolean,
+                                     var lifetimeByIdentifier: MutableMap<QueryID, QueryLifetime<Query>>)
+
+    private val state = AsyncSynchronized(State<QueryID, Query>(false, mutableMapOf()))
+
+
+    /**
+     * When a new QueryID is provided in the queries parameter and was not previously present in the state.lifeTimeByIdentifier,
+     * we start the effect with corresponding Query as initial value.
+     * The effect will remain active and generated events will be pushed back to the system
+     * as long as the same QueryID that started the effect is still present in subsequent calls.
+     *
+     *
+     * When for the same QueryID, a new distinct value of Query is provided, the already started effect
+     * will receive the new Query.
+     *
+     * When a QueryID that is present in state.lifeTimeByIdentifier but is not present in the provided
+     * queries the corresponding effect is disposed.
+     *
+     * @param queries used to activate, deactivate, update effects.
+     */
+    fun forwardQueries(queries: Map<QueryID, Query>) {
+        this.state.async { state ->
+            if (state.isDisposed) {
+                return@async
+            }
+            val lifetimeToUnsubscribeByIdentifier = state.lifetimeByIdentifier.toMutableMap()
+            for ((queryID, query) in queries) {
+                val queryLifetime = state.lifetimeByIdentifier[queryID]
+                if (queryLifetime != null) {
+                    lifetimeToUnsubscribeByIdentifier.remove(queryID)
+                    if (queryLifetime.latestQuery.value != query) {
+                        queryLifetime.latestQuery.onNext(query)
+                    } else continue
+                } else {
+                    val subscription = CompositeDisposable() // SingleAssignmentDisposable
```

Comment:
Yeah you are right... I was not very convinced of what I was writing there...

Reply:
>Also, does the `AsyncSynchronized` concept have a name? Is it used somewere else as well? I'd like to read more about it.

Thanks for the feedback... Maybe `SerialQueue`? I don't think there is a standard definition for  for this concept because the mutations can run on any thread vs serial execution queues which is a know concept on iOS. With serial execution queues all mutations are dispatched to the same thread. Not sure if @kzaher has any link to share with us...

Reply:
@markotron , I think it's called [AyncLock](https://github.com/atsushieno/rx/blob/master/Rx.NET/System.Reactive.Core/Reactive/Concurrency/AsyncLock.cs)